File: N/503-next-greater-element-ii/main.go
package main

import "fmt"

func nextGreaterElements(nums []int) []int {
	if len(nums) == 0 {
		return nil
	}
	if len(nums) == 1 {
		return []int{-1}
	}
	ret := make([]int, len(nums))
	stack := []int{nums[0]}
	pos := []int{0}

	for i := 1; i < len(nums); i++ {
		if nums[i] > stack[len(stack)-1] {
			idx := len(stack) - 1
			for idx >= 0 && stack[idx] < nums[i] {
				ret[pos[idx]] = nums[i]
				idx--
			}
			stack = stack[0 : idx+1]
			pos = pos[0 : idx+1]
		}
		stack = append(stack, nums[i])
		pos = append(pos, i)
	}

	if len(stack) > 0 {
		for i := 0; i <= pos[0]; i++ {
			if nums[i] > stack[len(stack)-1] {
				idx := len(stack) - 1
				for idx >= 0 && stack[idx] < nums[i] {
					ret[pos[idx]] = nums[i]
					idx--
				}
				stack = stack[0 : idx+1]
				pos = pos[0 : idx+1]
			}
		}
		for i := 0; i < len(pos); i++ {
			ret[pos[i]] = -1
		}
	}
	return ret
}

func main() {
	fmt.Println("a")
}

File: N/503-next-greater-element-ii/main.rs
struct Solution {}

impl Solution {
  pub fn next_greater_elements(nums: Vec<i32>) -> Vec<i32> {
    let mut posBuf: Vec<usize> = Vec::new();

    let mut stack: Vec<i32> = Vec::new();
    let mut ret: Vec<i32> = vec![-1; nums.len()];
    for i in 0..nums.len() {
      if stack.len() == 0 || stack[stack.len() - 1] > nums[i] {
        stack.push(nums[i]);
        posBuf.push(i);
      } else {
        let mut offset: i32 = -1;
        for j in (0..stack.len()).rev() {
          if stack[j] < nums[i] {
            ret[posBuf[j]] = nums[i];
          } else {
            offset = j as i32;
            break;
          }
        }
        stack.resize(offset as usize + 1, 0);
        posBuf.resize(offset as usize + 1, 0);
        stack.push(nums[i]);
        posBuf.push(i);
      }
    }

    let mut i: usize = 0;
    let mut j: usize = stack.len() - 1;
    while i < nums.len() && j >= 0 {
      if stack[j] < nums[i] {
        ret[posBuf[j]] = nums[i];
        j -= 1;
      } else {
        i += 1;
      }
    }
    for m in 0..j + 1 {
      ret[posBuf[m]] = -1;
    }
    ret
  }
}

fn main() {
  println!("{}", Solution::next_greater_elements(vec![0; 5]));
}
